src/crimson/os/seastore/extent_pinboard.cc

```diff
@@ -0,0 +1,637 @@
+    stats.add(warm);
+  }
+
+  void remove(CachedExtent &extent) final {
+    auto s = get_2q_state(extent);
+    if (extent.primary_ref_list_hook.is_linked()) {
+      if (s == extent_2q_state_t::WarmIn) {
+	warm_in.remove(extent);
+      } else {
+	ceph_assert(s == extent_2q_state_t::Hot);
+	hot.remove(extent);
+      }
+    }
+  }
+
+  void move_to_top(
+    CachedExtent &extent,
+    const Transaction::src_t* p_src) final {
+    auto state = get_2q_state(extent);
+    if (extent.primary_ref_list_hook.is_linked()) {
+      if (state == extent_2q_state_t::Hot) {
+	hot.move_to_top(extent, p_src);
+      } else {
+	ceph_assert(state == extent_2q_state_t::WarmIn);
+	// warm_in is a FIFO queue, do nothing here
+      }
+    } else if (!is_logical_type(extent.get_type())) {
+      // put physical extents to hot queue directly
+      ceph_assert(state == extent_2q_state_t::Hot ||
+		  state == extent_2q_state_t::Fresh);
+      on_update_hot(extent.get_loaded_length());
+      hot.move_to_top(extent, p_src);
+      set_2q_state(extent, extent_2q_state_t::Hot);
+    } else { // not in two queue
+      ceph_assert(state == extent_2q_state_t::Fresh);
+      auto queue = &warm_in;
+      auto new_state = extent_2q_state_t::WarmIn;
+      auto lext = extent.cast<LogicalCachedExtent>();
+      if (warm_out.remove_if_contains(lext->get_laddr())) {
+	queue = &hot;
+	new_state = extent_2q_state_t::Hot;
+	on_update_hot(extent.get_loaded_length());
+      } else {
+	on_update_warm_out(extent.get_loaded_length());
+      }
+      queue->move_to_top(extent, p_src);
+      set_2q_state(extent, new_state);
+    }
+  }
+
+  void increase_cached_size(
+    CachedExtent &extent,
+    extent_len_t increased_length,
+    const Transaction::src_t* p_src) final {
+    if (extent.primary_ref_list_hook.is_linked()) {
+      auto state = get_2q_state(extent);
+      if (state == extent_2q_state_t::WarmIn) {
+	on_update_warm_out(increased_length);
+	warm_in.increase_cached_size(extent, increased_length, p_src);
+      } else {
+	ceph_assert(state == extent_2q_state_t::Hot);
+	on_update_hot(increased_length);
+	hot.increase_cached_size(extent, increased_length, p_src);
+      }
+    }
+  }
+
+  void clear() final {
+    warm_in.clear();
+    warm_out.clear();
+    hot.clear();
+  }
+
+  ~ExtentPinboardTwoQ() {
+    clear();
+  }
+private:
+  void on_update_hot(size_t size) {
+    hot.for_each_to_be_trimmed_extent(size, [](CachedExtent &extent) {
+      set_2q_state(extent, extent_2q_state_t::Fresh);
+    });
+  }
+  void on_update_warm_out(size_t size) {
+    warm_in.for_each_to_be_trimmed_extent(
+      size,
+      [this](CachedExtent &extent) {
+	ceph_assert(is_logical_type(extent.get_type()));
+	set_2q_state(extent, extent_2q_state_t::Fresh);
+	auto lext = extent.cast<LogicalCachedExtent>();
+	warm_out.add(
+	  lext->get_laddr(),
+	  lext->get_loaded_length(),
+	  lext->get_type());
+      });
+  }
+  ExtentQueue warm_in;
+  IndexedFifoQueue warm_out;
+  ExtentQueue hot;
```

Comment:
Can A1 queue deal with (distinguish) sequential partial reads inside the same large extent?

Reply:
The "mixed io workload" test setup is that first fill the image with bs=1M io_depth=64, then randrw with bs=1M.

The new test setup could be fill image with bs=4M and randrs with bs=512K to test this case.

Reply:
> At least it's not worse than LRU...

Yeah, I also agree that 2Q is a generally better algorithm (if implemented properly).

> Can A1 queue deal with (distinguish) sequential partial reads inside the same large extent?

I think it's commen that extent granularity isn't align with reads, and IIUC it can cause an extent to be touched multiple times and be promoted to the main buffer inappropriately.

Probably we can add a TODO comment to the class, and deal with it in the next step.